### PR TITLE
Add import/export functionality

### LINKED ISSUE #206: Export / Import data
Since the app doesn't rely on any backend, if a user looses its phone, all data is gone. To solve this issue, let's add a manual save mechanism.

In settings, add two row : 

- Export data : this should trigger a function that will grab all user emissions from the redux store and store it in a `.json` file (or similar) that the user can save where he wants on his phone.

- Import data : this will ask user to select the `.json` file previously exported in order to replace `emissions` in the redux state with the data in the file.

_NB1 : this is could be a large and complicated task, probably not suitable for beginners._
_NB2 : due to spam, I won't assigned this task to anyone without previous contribution to this repo but PR are accepted, just have a look if there is no PR up and if this issue is not assigned to anyone._

settings

### PR DESCRIPTION
**Pull request by iSasFTW**
✅ I have read the [contributing file](https://github.com/NMF-earth/nmf-app/blob/master/contributing.md)

## Summary

This change closes #206. It adds import/export functionality, that allows the user to export the entire Redux store to a backup file that the user can then save to the desired location using a share sheet. (e.g saving to cloud services or local storage)

I haven't written any tests for this change, so that would be a todo, along with internationalization.

## Changelog

Added import/export functionality for backing up user data.

## Demo

Screenshot_1
Screenshot _2

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,10 +39,13 @@
     "expo": "39.0.0",
     "expo-asset": "8.2.0",
     "expo-constants": "9.2.0",
+    "expo-document-picker": "~8.4.0",
```
Comment (PierreBresson) [inline package.json:42]: Please read contributing file : https://github.com/NMF-earth/nmf-app/blob/master/contributing.md => no ~ in package.json

File: app/screens/Settings/SettingsScreen.tsx [hunk 2]
```diff
@@ -32,6 +38,62 @@ const SettingsScreen = (props) => {
...
+              throw "notBackup";
+            }
+          } catch {
+            Alert.alert("Error importing backup");
+          }
+        }
+      },
+    },
+    {
+      title: "Export",
```
Comment (PierreBresson) [inline app/screens/Settings/SettingsScreen.tsx:85]: t("SETTINGS_SCREEN_EXPORT")

File: app/screens/Settings/SettingsScreen.tsx [hunk 3]
```diff
@@ -32,6 +38,62 @@ const SettingsScreen = (props) => {
       onPress: navigator.openNotifications,
       os: "ios",
     },
+    {
+      title: "Import",
```
Comment (PierreBresson) [inline app/screens/Settings/SettingsScreen.tsx:42]: t("SETTINGS_SCREEN_IMPORT")

File: app/screens/Settings/SettingsScreen.tsx [hunk 4]
```diff
@@ -32,6 +38,62 @@ const SettingsScreen = (props) => {
...
+      onPress: async () => {
+        const file = await DocumentPicker.getDocumentAsync({});
+        if (file.type !== "cancel") {
+          try {
+            const data = JSON.parse(
+              await FileSystem.readAsStringAsync(file.uri)
+            );
+            if (data._persist) {
+              store.dispatch(
```
Comment (PierreBresson) [inline app/screens/Settings/SettingsScreen.tsx:51]: let's only save emissions for now

File: app/screens/Settings/SettingsScreen.tsx [hunk 5]
```diff
@@ -32,6 +38,62 @@ const SettingsScreen = (props) => {
...
+                budget.actions.setMonthlyCarbonBudget(
+                  data.budget.monthlyCarbonBudget
+                )
+              );
+              store.dispatch(
+                emissions.actions.loadEmissionsFromImport(data.emissions)
```
Comment (PierreBresson) [inline app/screens/Settings/SettingsScreen.tsx:57]:
```
...
const SettingsScreen = (props) => {
  const navigator = navigate(props.navigation); 
  const dispatch = useDispatch(); // add this
...
```
and then here simply : `dispatch(emissions.actions.loadEmissionsFromImport(data));`

File: app/screens/Settings/SettingsScreen.tsx [hunk 6]
```diff
@@ -1,18 +1,24 @@
...
 import ExpoConstants from "expo-constants";
+import * as DocumentPicker from "expo-document-picker";
+import * as FileSystem from "expo-file-system";
+import * as Sharing from "expo-sharing";
 
 import * as WebBrowser from "expo-web-browser";
 
 import { ImagesAssets } from "constant";
 
+import store from "../../redux/store";
```
Comment (PierreBresson) [inline app/screens/Settings/SettingsScreen.tsx:19]: to be deleted

File: app/screens/Settings/SettingsScreen.tsx [hunk 7]
```diff
@@ -32,6 +38,62 @@ const SettingsScreen = (props) => {
...
+          } catch {
+            Alert.alert("Error importing backup");
+          }
+        }
+      },
+    },
+    {
+      title: "Export",
+      onPress: async () => {
+        const exportData = JSON.stringify(store.getState());
```
Comment (PierreBresson) [inline app/screens/Settings/SettingsScreen.tsx:87]: Let's save only emissions : let's use `useSelector` instead
```
const emissions = useSelector(emissions.selectors.getAllEmissions)
JSON.stringify(emissions)
```

**Review by PierreBresson — CHANGES_REQUESTED** [review]
Let's stick to emissions only export only like mentioned on the second point. Sorry I should have been more clear.
The PR is working well but is not following any convention used in the app. If you are willing to do the changes, let me know, I will take over.

**Comment by iSasFTW** [comment]
Ah, sorry about that, I must have somehow overlooked the part about only importing emissions.
I'll make the changes as soon as I can.

**Comment by PierreBresson** [comment]
@iSasFTW Due to #243 , `master` branch is now deleted and `main` is the new `master`, that's why the PR was closed and I can't reopen it, sorry :/

**Comment by iSasFTW** [comment]
@PierreBresson Should I make a new PR onto the new branch?

**Comment by PierreBresson** [comment]
@iSasFTW yes, against `main` branch

**Comment by PierreBresson** [comment]
@saskakol done in https://github.com/NMF-earth/nmf-app/pull/263, thanks for the guidance 🙂

**Pull request closed without merging**